kyo-core/shared/src/main/scala/kyo/chunks.scala

```diff
@@ -0,0 +1,464 @@
+                    @tailrec def loop(index: Int): Unit =
+                        if index < size then
+                            array(start + index) = seq(index + dropLeft)
+                            loop(index + 1)
+                    loop(0)
+        if !isEmpty then
+            loop(this, elements, 0, 0)
+    end copyTo
+
+    final def toArray(using ClassTag[T]): Array[T] =
+        val array = new Array[T](size)
+        copyTo(array, 0)
+        array
+    end toArray
+
+    final private def toArrayInternal: Array[T] =
+        this match
+            case c if c.isEmpty =>
+                cachedEmpty.array.asInstanceOf[Array[T]]
+            case c: Compact[T] =>
+                c.array
+            case c =>
+                c.toArray(using ClassTag(classOf[Any]))
+
+    override def equals(other: Any) =
+        (this eq other.asInstanceOf[Object]) || {
+            other match
+                case other: Chunk[?] =>
+                    (this.isEmpty && other.isEmpty) ||
+                    (this.size == other.size && {
+                        val s = this.size
+                        val a = this.toIndexed
+                        val b = other.toIndexed
+                        @tailrec def loop(idx: Int = 0): Boolean =
+                            if idx < s then
+                                inline given CanEqual[Any, Any] = CanEqual.derived
+                                a(idx) == b(idx) && loop(idx + 1)
+                            else
+                                true
+                        loop()
+                    })
+                case _ =>
+                    false
+        }
+
+    override def hashCode(): Int =
+        if isEmpty then
+            1
+        else
+            val s = this.size
+            val a = this.toIndexed
+            @tailrec def loop(idx: Int = 0, acc: Int = 1): Int =
+                if idx < s then
+                    loop(idx + 1, 31 * acc + a(idx).hashCode())
+                else
+                    acc
+            loop()
+end Chunk
+
+object Chunks:
+
+    import internal.*
+
+    sealed abstract class Indexed[T] extends Chunk[T]:
+
+        //////////////////
+        // O(1) methods //
+        //////////////////
+
+        def apply(i: Int): T
+
+        final def head: T =
+            if isEmpty then
+                throw new NoSuchElementException
+            else
+                apply(0)
+
+        final def tail: Indexed[T] =
+            if size <= 1 then cachedEmpty.asInstanceOf[Indexed[T]]
+            else
+                this match
+                    case Tail(chunk, offset, size) =>
+                        Tail(chunk, offset + 1, size - 1)
+                    case c =>
+                        Tail(c, 1, size - 1)
+    end Indexed
+
+    def init[T]: Chunk[T] =
+        cachedEmpty.asInstanceOf[Chunk[T]]
+
+    def init[T](values: T*): Chunk[T] =
+        initSeq(values)
+
+    def initSeq[T](values: Seq[T]): Chunk[T] =
```

Comment:
Will there be a collision if we overload with `init`?

Reply:
I think so. It becomes ambiguous in case you want to make a chunk of `Seq`s:
```scala
Chunk.init(Seq(1)) // should it be a Chunk[Int] or a Chunk[Seq[Int]]?
```